dataactbroker/handlers/fileHandler.py

```diff
@@ -642,6 +643,130 @@ def generate_detached_file(self, file_type, cgac_code, start, end):
+            except ValueError:
+                raise ResponseException("Date must be provided as DD/MM/YYYY", StatusCode.CLIENT_ERROR, ValueError)
+
+            # the front-end is doing date checks, but we'll also do a few server side to ensure everything is correct
+            # when clients call the API directly
+            if job_data.get('reporting_start_date') > job_data.get('reporting_end_date'):
+                raise ResponseException("Submission start date {} is after the end date {}".format(
+                        job_data.get('reporting_start_date'), job_data.get('reporting_end_date')),
+                        StatusCode.CLIENT_ERROR)
+
+            if not current_user_can('writer', job_data["cgac_code"]):
+                raise ResponseException("User does not have permission to create jobs for this agency",
+                                        StatusCode.PERMISSION_DENIED)
+
+            submission = create_submission(g.user.user_id, job_data, None)
+            sess.add(submission)
+            sess.commit()
+            sub_tier_affiliation = SubmissionSubTierAffiliation(submission_id=submission.submission_id,
+                                                                sub_tier_agency_id=sub_tier_agency.sub_tier_agency_id)
+            sess.add(sub_tier_affiliation)
+            sess.commit()
+
+            # build fileNameMap to be used in creating jobs
+            for file_type in ['detached_award']:
+                # if file_type not included in request, and this is an update to an existing submission, skip it
+                if not request_params.get(file_type):
+                    # this is a new submission, all files are required
+                    raise ResponseException("Must include all required files for new submission",
+                                            StatusCode.CLIENT_ERROR)
+
+                file_name = request_params.get(file_type)
+                if file_name:
+                    if not self.isLocal:
+                        upload_name = "{}/{}".format(
+                            g.user.user_id,
+                            S3UrlHandler.get_timestamped_filename(file_name)
+                        )
+                    else:
+                        upload_name = file_name
+
+                    response_dict[file_type+"_key"] = upload_name
+                    upload_files.append(FileHandler.UploadFile(
+                        file_type=file_type,
+                        upload_name=upload_name,
+                        file_name=file_name,
+                        file_letter=FILE_TYPE_DICT_LETTER[FILE_TYPE_DICT[file_type]]
+                    ))
+
+            file_job_dict = create_jobs(upload_files, submission, False)
+            for file_type in file_job_dict.keys():
+                if "submission_id" not in file_type:
+                    response_dict[file_type+"_id"] = file_job_dict[file_type]
+            if create_credentials and not self.isLocal:
```

Comment:
This response is the same as `submit` and so is the build fileNameMap stuff, right? Can we abstract it out to a method call if possible instead of duplicating the same code?